### PR TITLE
Introduce planks type

### PR DESCRIPTION
**Pull request by ShockedPlot7560**
## Introduction
With Mojang's introduction of mosaic bamboo, it's necessary to have a way of distinguishing it.  
This is a somewhat direct way of dividing the blocks inheriting the planks from those inheriting the wood, but it is the most flexible way of making it easier to introduce mosaic bamboo blocks.

This PR goes hand in hand with #5875, which seems to need this API.

### API changes
- Added ``PlanksType``, it inserts ``WoodType`` values as standard, but may be extended in the future.
- Remove ``getWoodType`` for using ``getPlanksType`` in ``Planks``, ``WoodenSlab`` and ``WoodenStairs``
- Added ``PlanksLikeBlockIdHelper`` 
- Rename ``WoodType`` to ``LogType``
- Rename ``WoodTypeTrait`` to ``LogTypeTrait``
- Change ``writeLegacyWoodType`` to ``writeLegacyLogType`` in ``BlockStateWriter``
- Rename ``getWoodType`` to ``getLogType`` in ``BoatType``

## Backwards compatibility
- Api changed in `Planks`, `WoodenSlab` and `WoodenStairs`
- ``WoodType`` renamed in ``LogType``
- ``WoodTypeTrait`` renamed in ``LogTypeTrait``
- Change ``writeLegacyWoodType`` to ``writeLegacyLogType`` in ``BlockStateWriter``

## Follow-up
- Implements Bamboo Mosaic

### CODE REVIEW

**Comment by dktapps** [comment]
WoodType should probably be scrapped and replaced with LogType

File: src/block/BaseSign.php [hunk 1]
```diff
@@ -57,8 +57,8 @@ abstract class BaseSign extends Transparent{
 	/**
 	 * @param \Closure() : Item $asItemCallback
 	 */
-	public function __construct(BlockIdentifier $idInfo, string $name, BlockTypeInfo $typeInfo, WoodType $woodType, \Closure $asItemCallback){
-		$this->woodType = $woodType;
+	public function __construct(BlockIdentifier $idInfo, string $name, BlockTypeInfo $typeInfo, LogType $logType, \Closure $asItemCallback){
```
Comment (dktapps) [inline src/block/BaseSign.php]: This doesn't seem right... there are bamboo signs

File: src/block/FenceGate.php [hunk 2]
```diff
@@ -36,7 +36,7 @@
 use pocketmine\world\sound\DoorSound;
 
 class FenceGate extends Transparent{
-	use WoodTypeTrait;
+	use LogTypeTrait;
```
Comment (dktapps) [inline src/block/FenceGate.php]: Same again here

File: src/block/PlanksLikeBlockIdHelper.php [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+			PlanksType::ACACIA()->id() => Ids::ACACIA_STAIRS,
+			PlanksType::DARK_OAK()->id() => Ids::DARK_OAK_STAIRS,
+			PlanksType::MANGROVE()->id() => Ids::MANGROVE_STAIRS,
+			PlanksType::CRIMSON()->id() => Ids::CRIMSON_STAIRS,
+			PlanksType::WARPED()->id() => Ids::WARPED_STAIRS,
+			PlanksType::CHERRY()->id() => Ids::CHERRY_STAIRS,
+			default => throw new AssumptionFailedError("All planks types should be covered")
+		});
+	}
+}
```
Comment (dktapps) [inline src/block/PlanksLikeBlockIdHelper.php]: I'm not really sure what's gained by pulling this out into its own class.

File: src/block/WoodenDoor.php [hunk 4]
```diff
@@ -23,8 +23,8 @@
 
 namespace pocketmine\block;
 
-use pocketmine\block\utils\WoodTypeTrait;
+use pocketmine\block\utils\LogTypeTrait;
 
 class WoodenDoor extends Door{
-	use WoodTypeTrait;
+	use LogTypeTrait;
```
Comment (dktapps) [inline src/block/WoodenDoor.php]: Bamboo door is a thing

File: src/block/WoodenFence.php [hunk 5]
```diff
@@ -23,20 +23,20 @@
 
 namespace pocketmine\block;
 
-use pocketmine\block\utils\WoodTypeTrait;
+use pocketmine\block\utils\LogTypeTrait;
 
 class WoodenFence extends Fence{
-	use WoodTypeTrait;
+	use LogTypeTrait;
```
Comment (dktapps) [inline src/block/WoodenFence.php]: and bamboo fence

File: src/block/WoodenPressurePlate.php [hunk 6]
```diff
@@ -23,10 +23,10 @@
 
 namespace pocketmine\block;
 
-use pocketmine\block\utils\WoodTypeTrait;
+use pocketmine\block\utils\LogTypeTrait;
 
 class WoodenPressurePlate extends SimplePressurePlate{
-	use WoodTypeTrait;
+	use LogTypeTrait;
```
Comment (dktapps) [inline src/block/WoodenPressurePlate.php]: and pressure plate

File: src/block/WoodenTrapdoor.php [hunk 7]
```diff
@@ -23,10 +23,10 @@
 
 namespace pocketmine\block;
 
-use pocketmine\block\utils\WoodTypeTrait;
+use pocketmine\block\utils\LogTypeTrait;
 
 class WoodenTrapdoor extends Trapdoor{
-	use WoodTypeTrait;
+	use LogTypeTrait;
```
Comment (dktapps) [inline src/block/WoodenTrapdoor.php]: and trapdoor

File: src/block/utils/PlanksType.php [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+ */
+final class PlanksType{
+	use EnumTrait {
+		__construct as private Enum___construct;
+	}
+
+	protected static function setup() : void{
+		self::registerAll(
+			// TODO: bamboo mosaic
+			...self::fromLogType()
```
Comment (dktapps) [inline src/block/utils/PlanksType.php]: This is generally not recommended

File: src/item/BoatType.php [hunk 9]
```diff
@@ -47,26 +47,26 @@ final class BoatType{
...
-			new self("acacia", WoodType::ACACIA()),
-			new self("dark_oak", WoodType::DARK_OAK()),
-			new self("mangrove", WoodType::MANGROVE()),
+			new self("oak", LogType::OAK()),
+			new self("spruce", LogType::SPRUCE()),
+			new self("birch", LogType::BIRCH()),
+			new self("jungle", LogType::JUNGLE()),
+			new self("acacia", LogType::ACACIA()),
+			new self("dark_oak", LogType::DARK_OAK()),
+			new self("mangrove", LogType::MANGROVE()),
```
Comment (dktapps) [inline src/item/BoatType.php]: boats are made from planks

**Review by dktapps — CHANGES_REQUESTED** [review]
I think there is some confusion here over what's appropriate to use for the different types of blocks.

File: src/block/utils/LogTypeTrait.php [hunk 10]
```diff
@@ -26,15 +26,15 @@
 use pocketmine\block\BlockIdentifier;
 use pocketmine\block\BlockTypeInfo;
 
-trait WoodTypeTrait{
-	private WoodType $woodType; //immutable for now
+trait LogTypeTrait{
```
Comment (dktapps) [inline src/block/utils/LogTypeTrait.php:29]: Seems like this trait is now redundant, as far as I can see

File: src/block/utils/PlanksType.php [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+ * @method static PlanksType SPRUCE()
+ * @method static PlanksType WARPED()
+ */
+final class PlanksType{
+	use EnumTrait {
+		__construct as private Enum___construct;
+	}
+
+	protected static function setup() : void{
+		foreach(LogType::getAll() as $type) {
```
Comment (dktapps) [inline src/block/utils/PlanksType.php:51]: This is probably not a good idea - it won't translate well to native enums

**Comment by dktapps** [comment]
I feel like `PlanksType` should remain as `WoodType`. `LogType` is the outlier here, since it only applies to one block type. This would minimize unnecessary API breaks and reduce the diff size.

**Comment by ShockedPlot7560** [comment]
> I feel like `PlanksType` should remain as `WoodType`. `LogType` is the outlier here, since it only applies to one block type. This would minimize unnecessary API breaks and reduce the diff size.

However, the reason for separating is to be able to distinguish between planks and logs. The problem arises with bamboo and bamboo planks.

**Comment by dktapps** [comment]
Bamboo is still a wood type, it just doesn't have an associated log type.

As I said, I don't think the BC break of renaming all of this stuff to `Planks` is justified. Only logs really need to change IMO.

**Comment by dktapps** [comment]
Closing this due to lack of activity and conflicts. Introducing a `LogType` would be a lot simpler and require less work.

**Pull request closed without merging**